### PR TITLE
Added conditional bins

### PR DESCRIPTION
**Pull request by Dobios**

### CODE REVIEW

**Comment by schoeberl** [comment]
Most is reformatting. A little bit hard to see what really changed.

**Comment by tjarker** [comment]
The idea was to add more elaborate ways of collecting coverage. Rather than just working with ranges, a function which evaluates a custom condition is passed to the bin and the coverage collector uses it as a sample function. This would allow for an easy expression of more complicated relations of inputs, like "have all these inputs been equal" where the actual value is not interesting but rather how the inputs compare to each other.

File: src/main/scala/chiselverify/coverage/package.scala [hunk 1]
```diff
@@ -386,9 +369,15 @@ package object coverage {
...
+      * @param condition an extra condition that can be used to consider a hit
       */
-    case class Bins(name: String, range: Range) {
+    case class Bins(name: String, range: Range, condition: () => Boolean = () => true) {
         def ==(that: Bins): Boolean = (name == that.name) && (range.start == that.range.start) && (range.end == that.range.end)
+
+        def sample(portName: String, value: Int, coverageDB: CoverageDB): Unit =
+            if(range.contains(value) && condition()) coverageDB.addBinHit(portName, name, value)
+
+        def serialize: String = s"Bin( $name, $range ${ if(condition != (() => true)) condition })"
```
Comment (Dobios) [inline src/main/scala/chiselverify/coverage/package.scala:380]: @schoeberl The main changes are here, all I did was add the possibility to make a bin depend on an extra condition in order to consider a hit.

Everything else was simply reformatting @parzival3 's code to make it match the style used throughout the project.

**Comment by Dobios** [comment]
@tjarker 
> The idea was to add more elaborate ways of collecting coverage. Rather than just working with ranges, a function which evaluates a custom condition is passed to the bin and the coverage collector uses it as a sample function. This would allow for an easy expression of more complicated relations of inputs, like "have all these inputs been equal" where the actual value is not interesting but rather how the inputs compare to each other.

Ah ok I had misunderstood what you wanted. So would something like the following work:
```scala
//This is pseudoscala
case class CondPoint(name: String, portA: Data, portB: Data)

case class CondBin(binName: String, condition: (BigInt, BigInt) => Boolean) {
    def sample(pointA: Data, pointB: Data) = 
        if(condition(pointA.peek().litValue, pointB.peek().litValue)) addHit(binName)
}
```

**Comment by schoeberl** [comment]
> On 27 May, 2021, at 17:57, Tjark Petersen ***@***.***> wrot
> The idea was to add more elaborate ways of collecting coverage. Rather than just working with ranges, a function which evaluates a custom condition is passed to the bin and the coverage collector uses it as a sample function. This would allow for an easy expression of more complicated relations of inputs, like "have all these inputs been equal" where the actual value is not interesting but rather how the inputs compare to each other.
> 
> 

This is actually a great idea! Could this be something that is missing in UVM? Maybe something we can only do when a language supports functional programming?

Martin

**Pull request merged**